src/Workspaces/Remote/ServiceHub/CodeLensOOP/ReferenceCodeLensProvider.cs

```diff
@@ -0,0 +1,228 @@
+                                // file path
+                                new CodeLensDetailEntryField() { Text =   Descriptor.FilePath },
+                                // line number
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.LineNumber.ToString() },
+                                // column number
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.ColumnNumber.ToString() },
+                                // reference text
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.ReferenceLineText },
+                                // reference start
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.ReferenceStart.ToString() },
+                                // reference end
+                                new CodeLensDetailEntryField() { Text = (referenceLocationDescriptor.ReferenceStart + referenceLocationDescriptor.ReferenceLength).ToString() },
+                                // reference long description
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.LongDescription },
+                                // reference image id
+                                new CodeLensDetailEntryField() { ImageId = imageId }, 
+                                // text before reference 2
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.BeforeReferenceText2 },
+                                // text before reference 1
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.BeforeReferenceText1 },
+                                // text after reference 1
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.AfterReferenceText1 },
+                                // text after reference 2
+                                new CodeLensDetailEntryField() { Text = referenceLocationDescriptor.AfterReferenceText2 }
+                            },
+                        };
+                    }).ToList(),
+
+                    // use default behavior
+                    PaneNavigationCommands = null
+                };
+
+                return details;
+            }
+
+            public void Invalidate()
+            {
+                // fire and forget
+                // this get called from roslyn remote host
+                InvalidatedAsync?.Invoke(this, EventArgs.Empty);
+            }
+
+            public Task TrackChangesAsync(CancellationToken cancellationToken)
+            {
+                return _rpc.InvokeWithCancellationAsync(
+                    nameof(IRemoteCodeLensReferencesFromPrimaryWorkspaceService.SetCodeLensReferenceCallback),
+                    new object[] { _projectIdGuid, Descriptor.FilePath }, cancellationToken);
```

Comment:
can we have some comments as to what TrackChanges is supposed to do, and how this maps over to the OOP side?  It' snot clear to me from reading all of this.

Reply:
sure.